contracts/upgrader/src/contract.rs

```diff
@@ -0,0 +1,317 @@
+    ///     type Error = ContractError;
+    ///
+    ///     fn version(env: Env) -> String {
+    ///         String::from_str(&env, "0.2.0")
+    ///     }
+    ///
+    ///     fn upgrade(env: Env, new_wasm_hash: BytesN<32>) -> Result<(), ContractError> {
+    ///         Self::owner(&env).require_auth();
+    ///
+    ///         env.deployer().update_current_contract_wasm(new_wasm_hash);
+    ///         Ok(())
+    ///     }
+    /// }
+    ///
+    /// #[contractimpl]
+    /// impl DummyContract {
+    ///     pub fn migrate(env: Env, migration_data: String) {
+    ///         Self::owner(&env).require_auth();
+    ///
+    ///         env.storage()
+    ///             .instance()
+    ///             .set(&DataKey::Data, &migration_data);
+    ///     }
+    ///
+    ///     fn owner(env: &Env) -> Address {
+    ///         env.storage().instance().get(&DataKey::Owner).unwrap()
+    ///     }
+    /// }
+    const WASM_AFTER_UPGRADE: &[u8] = include_bytes!("testdata/dummy.wasm");
+
+    #[test]
+    fn upgrade_and_migrate_are_atomic() {
+        let env = Env::default();
+
+        let contract_address = env.register(DummyContract, ());
+        let upgrader_address = env.register(Upgrader, ());
+
+        let owner = set_owner(&env, &contract_address);
+
+        let original_version: String = query_version(&env, &contract_address);
+        assert_eq!(original_version, String::from_str(&env, "0.1.0"));
+
+        let hash_after_upgrade = env.deployer().upload_contract_wasm(WASM_AFTER_UPGRADE);
+        let expected_data = String::from_str(&env, "migration successful");
+        let expected_version = String::from_str(&env, "0.2.0");
+
+        let (upgrade_auth, migrate_auth) =
+            build_invocation_auths(&env, &contract_address, &hash_after_upgrade, &expected_data);
+
+        // add the owner to the set of authenticated addresses
+        env.mock_auths(&[
+            MockAuth {
+                address: &owner,
+                invoke: &upgrade_auth,
+            },
+            MockAuth {
+                address: &owner,
+                invoke: &migrate_auth,
+            },
+        ]);
+
+        UpgraderClient::new(&env, &upgrader_address).upgrade(
+            &contract_address,
+            &expected_version,
+            &hash_after_upgrade,
+            &soroban_sdk::vec![&env, expected_data.to_val()],
+        );
+
+        // ensure new version is set correctly
+        let upgraded_version: String = env.invoke_contract(
+            &contract_address,
+            &Symbol::new(&env, VERSION),
+            soroban_sdk::vec![&env],
+        );
+        assert_eq!(upgraded_version, expected_version);
+
+        // ensure migration was successful
+        env.as_contract(&contract_address, || {
+            let data: String = env.storage().instance().get(&DataKey::Data).unwrap();
+            assert_eq!(data, expected_data);
+        });
+    }
+
+    #[test]
+    #[should_panic]
```

Comment:
It would be good to specify the expected panic reason
```suggestion
    #[should_panic(expected = "HostError: Error(Auth, InvalidAction)")]
```